src/System.Private.CoreLib/shared/System/Collections/Generic/List.cs

```diff
@@ -469,7 +469,7 @@ public int FindIndex(int startIndex, int count, Predicate<T> match)
                 ThrowHelper.ThrowStartIndexArgumentOutOfRange_ArgumentOutOfRange_Index();
             }
 
-            if (count < 0 || startIndex > _size - count)
+            if ((uint)count > (uint)(_size - startIndex))
```

Comment:
FWIW this is one case that may be worth doing, provided that there's measurable perf improvement. There's little chance for the JIT to prove that `_size - startIndex` is positive, even if looking at the previous `(uint)startIndex > (uint)_size` test. That's because the previous test assumes that `_size` is positive but proving that in the JIT would require interprocedural analysis.

Reply:
How could the JIT assume that `_size` is positive in the previous test?

Reply:
That's what I was saying - the JIT can't assume that, it needs to prove it and that's difficult. The assumption was done by however wrote that `uint` test.

Reply:
> Does the JIT only optimize built-in collection types?

If the JIT can't know whether `_size` is positive or not, I doubt that it can optimize the expression `index < 0 || index > _size` to any degree.

Or do you plan to add some attributes to non-negative fields like `_size`?